Add guard in ruma_api! macro for non-ASCII in path

DevinR528 opened the pull request:
Returns a compiler error if any non-ASCII characters are found. Add trybuild test for invalid path characters.

farodin91 commented:
@jimmycuadra Review?

farodin91 commented:
I need to add an event `m.tag`.

farodin91 commented:
This could done by different PR. It need to fix ruma/ruma-events#4, before this issue could resolve correct.

mujx commented on src/api/r0/tags.rs:
```diff
@@ -0,0 +1,284 @@
+
+        // Check if the given user_id corresponds to the authenticated user.
+        if user_id != user.id {
+            Err(ApiError::unauthorized(None))?;
```
Could you add some useful error message instead of `None`? You can use [this](https://github.com/mujx/ruma/blob/0bcceec9bed88b26ee0500b1483a73098ca0c850/src/api/r0/account.rs#L153) for this particular error, in order to be consistent.

mujx commented on src/tags.rs:
```diff
@@ -0,0 +1,155 @@
+        room_id: RoomId,
+        tag: String,
+        content: Value)
+        -> Result<(), ApiError> {
```
Until we use something like `rustfmt` it would be better to stick with the [current style](https://github.com/ruma/ruma/blob/9e9bdc2801223937d7b85a7b57a2db6be23f6ee2/src/room.rs#L99).

farodin91 replied:
Resolved

mujx commented on src/api/r0/tags.rs:
```diff
@@ -0,0 +1,284 @@
+
+        let mut content = String::new();
+        if let Err(_) = request.body.read_to_string(&mut content) {
+            Err(ApiError::not_found(None))?;
```
You missed this one. Although synapse [parses](https://github.com/matrix-org/synapse/blob/e9c1cabac26cf8e28152ebdb3caf29d4457eea0e/synapse/rest/client/v2_alpha/tags.py#L73) it as json, so you can skip this check and just add the `JsonRequest` middleware.

farodin91 replied:
Resolved!
I think the json middleware could be refactored in the future.

mujx commented on src/api/r0/tags.rs:
```diff
@@ -0,0 +1,284 @@
+        let user = request.extensions.get::<User>()
+            .expect("AccessTokenAuth should ensure a user").clone();
+        let params = request.extensions.get::<Router>().expect("Params object is missing").clone();
+        let tag = match params.find("tag") {
```
You can implement this check as a middleware (e.g `TagParam`).

farodin91 replied:
Resolved!

farodin91 commented:
Waiting for #114

farodin91 commented:
@jimmycuadra Could you start reviewing? Thanks

mujx commented on src/models/tags.rs:
```diff
@@ -0,0 +1,166 @@
+        delete(tag)
+            .execute(connection)
+            .map_err(|err| match err {
+                DieselError::NotFound => ApiError::not_found(None),
```
Also a missing error message here.

mujx commented on src/models/tags.rs:
```diff
@@ -0,0 +1,166 @@
+        user_id: UserId,
+        room_id: RoomId,
+        tag: String,
+        content: Value,
```
Why `Value` instead of `String`?

mujx commented on src/api/r0/tags.rs:
```diff
@@ -0,0 +1,224 @@
+    }
+
+    #[test]
+    fn update_tag() {
```
You can add another test where you update a non-existent tag.

farodin91 replied:
No, See Line 177, 179

mujx replied:
Sorry, my bad. Rename the helper function to `put_tag` so there is no confusion.

mujx commented on src/api/r0/tags.rs:
```diff
@@ -0,0 +1,224 @@
+    fn handle(&self, request: &mut Request) -> IronResult<Response> {
+        let user_id = request.extensions.get::<UserIdParam>()
+            .expect("UserIdParam should ensure a UserId").clone();
+        let room_id = request.extensions.get::<RoomIdParam>()
```
You need to check that the room and user the exist before adding a tag. Also add the relevant tests. We don't have foreign keys so there will be no error.

farodin91 replied:
Solved for all three, but not nice for delete.

mujx commented on src/models/tags.rs:
```diff
@@ -0,0 +1,166 @@
+            map.insert(tag.tag, info);
+        }
+
+        let tags = TagEventContent { tags: map };
```
Better use a custom `TagsResponse` in the api handler and here just return the hashmap.

farodin91 commented on src/api/r0/tags.rs:
```diff
@@ -0,0 +1,224 @@
+        let room_id = test.create_public_room(&access_token);
+        let put_tag_path = format!(
+            "/_matrix/client/r0/user/{}/rooms/{}/tags/{}?access_token={}",
+            "@carls:ruma.test",
```
I would do this after is fixed #121.

jimmycuadra commented:
Fantastic work. :D

DevinR528 committed:
Add guard in ruma_api! macro for non-ASCII in path

Returns a compiler error if any non ASCII characters are found. Add
trybuild test for invalid path characters.

jplatte commented on ruma-api-macros/src/api.rs:
```diff
@@ -269,8 +269,8 @@ impl ToTokens for Api {
-                        // The only way this can fail is if the path given in the API definition is
-                        // invalid. It is okay to panic in that case.
+                        // The ruma_api! macro guards against invalid path input but, if there are
+                        // invalid (non ASCII) bytes in the fields with the query attribute this will panic.
```
This comment makes it sound like this is *not* guaranteed to succeed.

DevinR528 replied:
Won't it fail if the query string had `vec![("abc", "µ / ° / § / €")]` it would fail? Or is it impossible for that to happen because of percent encoding?

jplatte replied:
Hm, I don't actually know. I now understand what this comment is talking about though :D

Just put the comma before the "but" and I'm happy for now :)

jplatte commented on ruma-api-macros/src/api/metadata.rs:
```diff
@@ -61,6 +61,12 @@ impl TryFrom<RawMetadata> for Metadata {
+                        if !literal.value().is_ascii() {
+                            return Err(syn::Error::new_spanned(
+                                literal,
+                                "path may only contain printable ASCII characters",
```
`is_ascii` checks whether characters are ASCII, not whether they're *printable* ASCII. The latter is probably the check we want though.

jplatte commented on ruma-api/tests/ui/02-invalid-path.rs:
```diff
@@ -0,0 +1,21 @@
+        description: "Does something.",
+        method: GET,
+        name: "newtype_body_endpoint",
+        path: "/_matrix/some/query/map/\u{ffff}",
```
Maybe we could use an example that doesn't require `\u{}`? Something simpler like µ / ° / § / € (al of which can easily be typed on a German keyboard, not sure about others).

DevinR528 committed:
Use is_printable over is_ascii for invalid path char

DevinR528 committed:
Fix clippy into_iter warning

jplatte commented on ruma-api-macros/src/util.rs:
```diff
@@ -257,3 +257,11 @@ pub(crate) fn req_res_name_value<T>(
 }
+
+fn is_ascii_printable(b: &u8) -> bool {
+    b.wrapping_sub(b' ') < 0x5F
```
That's a weird check, I can't tell if it is correct. Does this do the same thing as `(0x20..=0x7E).contains(b)`?

DevinR528 replied:
That works the same, I changed it to the above as that's way more readable the one I found was from [ascii](https://docs.rs/ascii/1.0.0/src/ascii/ascii_char.rs.html#537-539). That's cool you can use the range like that I didn't know!

Is space " " an invalid character for paths?

jplatte replied:
Now that you mention it, that is considered printable, huh? Yeah, let's forbid it for now. It would have to be percent-encoded at the very least, but I don't see Matrix ever adding endpoints with spaces in their paths :D

jplatte commented on ruma-api-macros/src/util.rs:
```diff
@@ -257,3 +257,11 @@ pub(crate) fn req_res_name_value<T>(
+    b.wrapping_sub(b' ') < 0x5F
+}
+
+pub(crate) fn is_printable(string: &str) -> bool {
```
`is_printable` is very misleading if it actually checks whether all characters are printable *ASCII* chars. I'd rename this `is_ascii_printable` and inline the helper function above.

jplatte commented on ruma-api/tests/ui/02-invalid-path.rs:
```diff
@@ -5,7 +5,7 @@ ruma_api! {
         method: GET,
         name: "newtype_body_endpoint",
-        path: "/_matrix/some/query/map/\u{ffff}",
+        path: "µ / ° / § / €",
```
Haha, I actually meant including just one characters like that instead of `\u{ffff}`, but this works too ^^

DevinR528 replied:
It's a really invalid path :laughing:

DevinR528 committed:
Rename is_printable -> is_ascii_printable

DevinR528 committed:
Return error if ruma_api! macro finds spaces in metadata's path field

jplatte commented on ruma-api-macros/src/api/metadata.rs:
```diff
@@ -61,6 +61,13 @@ impl TryFrom<RawMetadata> for Metadata {
                 "path" => match expr {
                     Expr::Lit(ExprLit { lit: Lit::Str(literal), .. }) => {
+                        let path_str = literal.value();
+                        if !util::is_ascii_printable(&path_str) || path_str.contains(' ') {
```
We could just update `is_ascii_printable` to check `0x21` onwards instead of `0x20` and rename it to sth. like `is_valid_endpoint_path`.

DevinR528 replied:
Yeah I like that I just couldn't think of a good name to call it but `is_valid_endpoint_path` that is perfect.

DevinR528 committed:
Rename is_ascii_printable -> is_valid_endpoint_path and move space check in

jplatte commented:
Thanks! Will merge once CI is happy.
